### PR TITLE
[ApiCompat] Refactor of the task

### PR DESCRIPTION
**Pull request by Aliandi**
- Add class to handle command line results
- Add class to handle ApiCompat commands
- Delegate responsabilities to those classes
- Clean up index.ts

### CODE REVIEW

File: ApiCompatTask/source/commandLineResult.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    private getTotalIssues = (message: string, indexOfTotalIssues: number): number => {
+        return parseInt(message.substring(indexOfTotalIssues).split(':')[1].trim(), 10);
+    }
+    
+    private getBody = (message: string, indexOfTotalIssues: number): string => {
+        return message.substring(0, indexOfTotalIssues - 1);
+    }
+
+    public resultText() {
+        return this.totalIssues?
```
Comment (denscollo) [inline ApiCompatTask/source/commandLineResult.ts]: Add space between totalIssues and the  ?. Right now is like this `this.totalIssues?` and it'd be `this.totalIssues ?`

File: ApiCompatTask/source/commandLineResult.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        `There were differences between the assemblies` :
+        `No differences were found between the assemblies` ;
```
Comment (denscollo) [inline ApiCompatTask/source/commandLineResult.ts]: Remove the space before the `;`

File: ApiCompatTask/source/commandLineResult.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    public compattibilityResult = (): TaskResult => {
+        return this._totalIssues ?
+            TaskResult.Succeeded : getInput('failOnIssue') ?
+            TaskResult.Failed : TaskResult.SucceededWithIssues;
+    }
+
+    public colorCode = (): string => {
+        return this._totalIssues === 0
```
Comment (denscollo) [inline ApiCompatTask/source/commandLineResult.ts]: I think we'd refactor this a little bit. There're too many ternaries nested

**Comment by ParadoxARG** [comment]
I personally think that the pipelines task use inline methods for performance. We need to keep in mind that Azure charges for each second it takes. I'm more inclined to have it that way, and that's the current way that it's implemented in all the [currently available tasks](https://github.com/Microsoft/azure-pipelines-tasks).

File: ApiCompatTask/source/commandLineResult.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+            ? "\x1b[32m"
```
Comment (GasparAcevedoZainSouthworks) [inline ApiCompatTask/source/commandLineResult.ts]: Can we add a comment describing each situation/color?

File: ApiCompatTask/source/commandLineResult.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    constructor( result: string ) {
+        const indexOfTotalIssues = result.indexOf("Total Issues");
+        this._totalIssues = this.getTotalIssues(result, indexOfTotalIssues);
+        this._body = this.getBody(result, indexOfTotalIssues);
+    }
+
+    private getTotalIssues = (message: string, indexOfTotalIssues: number): number => {
+        return parseInt(message.substring(indexOfTotalIssues).split(':')[1].trim(), 10);
```
Comment (GasparAcevedoZainSouthworks) [inline ApiCompatTask/source/commandLineResult.ts]: Is there any case where this `split` method returns `null` or only one item?

Comment (Aliandi) [reply ApiCompatTask/source/commandLineResult.ts]: Nope! the task will always return 

> Total Issues : NUMBER

**Comment by ParadoxARG** [comment]
> I personally think that the pipelines task use inline methods for performance. We need to keep in mind that Azure charges for each second. I'm more inclined to have it that way, and that's the current way that it's implemented in all the [currently available tasks](https://github.com/Microsoft/azure-pipelines-tasks).

I thought about this one and I changed my mind. The impact in performance should be really small and we should go for more maintainable code.

**Comment by Aliandi** [comment]
> 
> 
> > I personally think that the pipelines task use inline methods for performance. We need to keep in mind that Azure charges for each second. I'm more inclined to have it that way, and that's the current way that it's implemented in all the [currently available tasks](https://github.com/Microsoft/azure-pipelines-tasks).
> 
> I thought about this one and I changed my mind. The impact in performance should be really small and we should go for more maintainable code.

I agree. We could run some performance tests just to be sure, but according to what I have been reading we shouldn't have issues.

File: ApiCompatTask/source/commandLineResult.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+    get totalIssues() {
+        return  this._totalIssues;
+    }
+
+    get body() {
+        return this._body;
+    }
+
+    constructor( result: string ) {
```
Comment (GasparAcevedoZainSouthworks) [inline ApiCompatTask/source/commandLineResult.ts]: The constructor should be `constructor(result: string)`, I.E, without the empty spaces after `(` and before `)`

**Pull request merged**